app/models/user.rb

```diff
@@ -43,6 +48,21 @@ def self.from_omniauth(auth)
     user
   end
 
+  def self.authenticate(email, password)
+    user = find_by_email email
+    if user && user.password_hash == BCrypt::Engine.hash_secret(password, user.password_salt)
+      user
+    else
+      nil
+    end
+  end
+
+  def hash_password
+    return unless password.present?
+    self.password_salt = BCrypt::Engine.generate_salt
+    self.password_hash = BCrypt::Engine.hash_secret(password, password_salt)
```

Comment:
This is probably secure, but I'd suggest using the Rails built in facility: http://api.rubyonrails.org/classes/ActiveModel/SecurePassword/ClassMethods.html#method-i-has_secure_password

Unrelated, but in searching I also found Rails now has a secure token facility too:  https://github.com/rails/rails/blob/master/activerecord/lib/active_record/secure_token.rb